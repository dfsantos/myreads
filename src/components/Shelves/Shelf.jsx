import React from 'react';
import PropTypes from 'prop-types';
import Paper from 'material-ui/Paper';

import Book from '../Book';

const propTypes = {
  books: PropTypes.array,
  name: PropTypes.string.isRequired,
  shelf: PropTypes.string.isRequired,
  onCategorizeBook: PropTypes.func.isRequired,
};

const defaultProps = {
  books: [],
};

const byShelf = shelf => book => book.shelf === shelf;

function Shelf({ name, shelf, books, onCategorizeBook }) {
  const shelfBooks = books.filter(byShelf(shelf));
  const isEmpty = shelfBooks.length === 0;
  return (
    <Paper className="shelf">
      <div className="shelf-title">
        <span>{name}</span>
      </div>
      <div className="shelf-content">
        {isEmpty && (
          <div className="empty-shelf-message">
            <span>This shelf is empty.</span>
          </div>
        )}
        {!isEmpty &&
          shelfBooks.map(book => {
            const { thumbnail } = book.imageLinks || { thumbnail: '' };
            return (
              <Book
                key={book.id}
                title={book.title}
                subtitle={book.subtitle}
                coverLink={thumbnail}
                authors={book.authors}
                description={book.description}
                shelf={book.shelf}
                onChangeBookShelf={onCategorizeBook(book)}
              />
            );
          })}
      </div>
    </Paper>
  );
}

Shelf.propTypes = propTypes;
Shelf.defaultProps = defaultProps;

export default Shelf;
