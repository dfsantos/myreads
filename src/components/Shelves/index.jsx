import React from 'react';
import PropTypes from 'prop-types';

import Shelf from './Shelf';

const propTypes = {
  books: PropTypes.array,
  onCategorizeBook: PropTypes.func.isRequired,
  configuration: PropTypes.array.isRequired,
};

const defaultProps = {
  books: [],
};

function Shelves({ books, configuration, onCategorizeBook }) {
  return (
    <div className="list-books">
      <div className="list-books-content">
        {configuration.map(shelf => (
          <Shelf
            key={shelf.type}
            name={shelf.name}
            shelf={shelf.type}
            books={books}
            onCategorizeBook={onCategorizeBook}
          />
        ))}
      </div>
    </div>
  );
}

Shelves.propTypes = propTypes;
Shelves.defaultProps = defaultProps;

export default Shelves;
