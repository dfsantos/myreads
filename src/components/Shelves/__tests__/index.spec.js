import React from 'react';
import toJson from 'enzyme-to-json';

import Shelves from '../index';

import configuration from '../../../config/shelf.config.json';

const onCategorizeBookListener = jest.fn(() => () => {});
const books = [{ title: 'Book A' }, { title: 'Book B' }];

it('should render without errors', () => {
  shallow(
    <Shelves
      books={books}
      configuration={configuration}
      onCategorizeBook={onCategorizeBookListener}
    />
  );
});

it('snapshot without books', () => {
  const wrapper = shallow(
    <Shelves
      books={books}
      configuration={configuration}
      onCategorizeBook={onCategorizeBookListener}
    />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('snapshot with books', () => {
  const wrapper = shallow(
    <Shelves
      books={books}
      configuration={configuration}
      onCategorizeBook={onCategorizeBookListener}
    />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});
