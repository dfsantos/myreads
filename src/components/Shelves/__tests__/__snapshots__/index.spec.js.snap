// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot with books 1`] = `
<div
  className="list-books"
>
  <div
    className="list-books-content"
  >
    <Shelf
      books={
        Array [
          Object {
            "title": "Book A",
          },
          Object {
            "title": "Book B",
          },
        ]
      }
      key="currentlyReading"
      name="Currently Reading"
      onCategorizeBook={[Function]}
      shelf="currentlyReading"
    />
    <Shelf
      books={
        Array [
          Object {
            "title": "Book A",
          },
          Object {
            "title": "Book B",
          },
        ]
      }
      key="read"
      name="Read"
      onCategorizeBook={[Function]}
      shelf="read"
    />
    <Shelf
      books={
        Array [
          Object {
            "title": "Book A",
          },
          Object {
            "title": "Book B",
          },
        ]
      }
      key="wantToRead"
      name="Want to Read"
      onCategorizeBook={[Function]}
      shelf="wantToRead"
    />
  </div>
</div>
`;

exports[`snapshot without books 1`] = `
<div
  className="list-books"
>
  <div
    className="list-books-content"
  >
    <Shelf
      books={
        Array [
          Object {
            "title": "Book A",
          },
          Object {
            "title": "Book B",
          },
        ]
      }
      key="currentlyReading"
      name="Currently Reading"
      onCategorizeBook={[Function]}
      shelf="currentlyReading"
    />
    <Shelf
      books={
        Array [
          Object {
            "title": "Book A",
          },
          Object {
            "title": "Book B",
          },
        ]
      }
      key="read"
      name="Read"
      onCategorizeBook={[Function]}
      shelf="read"
    />
    <Shelf
      books={
        Array [
          Object {
            "title": "Book A",
          },
          Object {
            "title": "Book B",
          },
        ]
      }
      key="wantToRead"
      name="Want to Read"
      onCategorizeBook={[Function]}
      shelf="wantToRead"
    />
  </div>
</div>
`;
